libp2p-swarm/src/transport.rs

```diff
@@ -754,93 +746,31 @@ where
 		Ok(Box::new(future))
 	}
 
-	/// Tries to dial on the `Multiaddr` using the transport that was passed to `new`, then upgrade
-	/// the connection. Also listens to incoming substream requires on that dialed connection, and
-	/// automatically upgrades the incoming substreams.
-	///
-	/// Note that this does the same as `MuxedTransport::dial_and_listen`, but with less
-	/// restrictions on the trait requirements.
-	pub fn dial_and_listen(
-		self,
-		addr: Multiaddr,
-	) -> Result<
-		Box<
-			Future<
-				Item = (
-					Box<Stream<Item = C::Output, Error = IoError> + 'a>,
-					Box<Future<Item = C::Output, Error = IoError> + 'a>,
-				),
-				Error = IoError,
-			>
-				+ 'a,
-		>,
-		(Self, Multiaddr),
-	>
-	where
-		T: MuxedTransport,
-		C::NamesIter: Clone, // TODO: not elegant
-		C: Clone,
+	/// If the underlying transport is a `MuxedTransport`, then after calling `dial` we may receive
+	/// substreams opened by the dialed nodes.
+	/// 
+	/// This function returns the next incoming substream. You are strongly encouraged to call it
+	/// if you have a muxed transport.
+	pub fn next_incoming(self) -> Box<Future<Item = (C::Output, Multiaddr), Error = IoError> + 'a>
```

Comment:
We can do this using `as fn(_) -> _` and then copying the giant resulting type signature out of the error message, although it's your call as to whether this is worth doing.

Reply:
We can't because unfortunately one of the closures captures a value.